### PR TITLE
fix ingress for k8s v1.22+

### LINKED ISSUE #12299: Fix Ingress on New Kubenretes version 1.22.0
as seen failure in this PR

https://github.com/kubernetes/minikube/pull/12283

https://storage.googleapis.com/minikube-builds/logs/12283/20054/Docker_Linux.html

### PR DESCRIPTION
**Pull request by prezha**
fixes #12299

tldr: with this pr, ingress should also work with the latest k8s v1.22+

there are still a lot of moving parts there and i'll try to document them here

background - deprecation announcement: [Updating NGINX-Ingress to use the stable Ingress API](https://kubernetes.io/blog/2021/07/26/update-with-ingress-nginx/)
  => Drop v1beta1 from ingress nginx (https://github.com/kubernetes/ingress-nginx/pull/7156)
    => Things to do before the v1 version is released (https://github.com/kubernetes/ingress-nginx/issues/7398)
          => docs for migration of ingress resources in K8s v1.22 and beyond (https://github.com/kubernetes/ingress-nginx/pull/7441)

#### NGINX Ingress Controller (aka _ingress-nginx/controller_)
_latest stable release v0.49.0 does not work with k8s v1.22+, v1.0.0 Beta 3 works but only with k8s v1.19+_ (https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v1.0.0-beta.3) introducing breaking changes: https://github.com/kubernetes/ingress-nginx/#support-versions-table

#### Kubernetes webhook certificate generator and patcher (aka _kube-webhook-certgen_)
so far, `jet/kube-webhook-certgen` was used (https://github.com/jet/kube-webhook-certgen), but it's not yet updated to work with the latest ingress-nginx/controller version, ie,
_docker.io/jettech/kube-webhook-certgen:v1.5.1 does not work, neither the newer docker.io/jettech/kube-webhook-certgen:v1.5.2_ (https://hub.docker.com/layers/jettech/kube-webhook-certgen/v1.5.2/images/sha256-4709d4110f667ba19875d17a23c04ab016c03ae6493d456fb41d942854b23ac1?context=explore)

there's an open issue (https://github.com/kubernetes/ingress-nginx/issues/7418) and pr (https://github.com/jet/kube-webhook-certgen/issues/30) that could address that at some point

in the meantime, upstream already `forked` this repo and adapted it (https://github.com/kubernetes/ingress-nginx/tree/main/images/kube-webhook-certgen):

> We moved it here so we can change / update the Kubernetes APIs, and we are really thankful to the original creators.

but https://github.com/kubernetes/ingress-nginx/tree/main/images states:

> Only the nginx image is meant to be published. Other images are used as examples for a feature of the ingress controller or to run e2e tests

so, i manually built the `kube-webhook-certgen` image from upstream repo (https://github.com/kubernetes/ingress-nginx/blob/main/images/kube-webhook-certgen/Makefile => https://hub.docker.com/repository/docker/prezha/kube-webhook-certgen), but then i've found the "official" image that can be used:
https://github.com/kubernetes/ingress-nginx/blob/557604f4ef526f7755d36089b617bc7686c389f9/deploy/static/provider/kind/deploy.yaml#L612
that is:
https://github.com/kubernetes/ingress-nginx/blob/557604f4ef526f7755d36089b617bc7686c389f9/deploy/static/provider/kind/deploy.yaml#L660
`k8s.gcr.io/ingress-nginx/kube-webhook-certgen:v1.0@sha256:f3b6b39a6062328c095337b4cadcefd1612348fdd5190b1dcbcb9b9e90bd8068`
##### just-ingress@docker test:

❯ env TEST_ARGS="-minikube-start-args='--driver=docker' -test.run TestAddons --cleanup=false" make integration
go test -ldflags="-X k8s.io/minikube/pkg/version.version=v1.22.0 -X k8s.io/minikube/pkg/version.isoVersion=v1.22.0-1628974786-12268 -X k8s.io/minikube/pkg/version.gitCommitID="69d5d34eeb4d5f3ad963c544773dd088d6550821-dirty" -X k8s.io/minikube/pkg/version.storageProvisionerVersion=v5" -v -test.timeout=90m ./test/integration --tags="integration " -minikube-start-args='--driver=docker' -test.run TestAddons --cleanup=false 2>&1 | tee "./out/testout_69d5d34ee.txt"
Found 16 cores, limiting parallelism with --test.parallel=9
=== RUN   TestAddons
=== RUN   TestAddons/Setup
    addons_test.go:80: (dbg) Run:  out/minikube start -p addons-20210822033320-3217571 --wait=true --memory=4000 --alsologtostderr --driver=docker --addons=ingress
    addons_test.go:80: (dbg) Done: out/minikube start -p addons-20210822033320-3217571 --wait=true --memory=4000 --alsologtostderr --driver=docker --addons=ingress: (1m7.078556595s)
=== RUN   TestAddons/parallel
=== RUN   TestAddons/parallel/Ingress
=== PAUSE TestAddons/parallel/Ingress
=== CONT  TestAddons/parallel/Ingress
    addons_test.go:170: (dbg) Run:  kubectl --context addons-20210822033320-3217571 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s
    addons_test.go:170: (dbg) Done: kubectl --context addons-20210822033320-3217571 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s: (13.875082612s)
    addons_test.go:177: (dbg) Run:  kubectl --context addons-20210822033320-3217571 replace --force -f testdata/nginx-ingv1.yaml
    addons_test.go:190: (dbg) Run:  kubectl --context addons-20210822033320-3217571 replace --force -f testdata/nginx-pod-svc.yaml
    addons_test.go:195: (dbg) TestAddons/parallel/Ingress: waiting 4m0s for pods matching "run=nginx" in namespace "default" ...
    helpers_test.go:343: "nginx" [a150e166-bb61-4d0c-a5d7-bb5d88311075] Pending / Ready:ContainersNotReady (containers with unready status: [nginx]) / ContainersReady:ContainersNotReady (containers with unready status: [nginx])
    helpers_test.go:343: "nginx" [a150e166-bb61-4d0c-a5d7-bb5d88311075] Running
    addons_test.go:195: (dbg) TestAddons/parallel/Ingress: run=nginx healthy within 12.015628991s
    addons_test.go:215: (dbg) Run:  out/minikube -p addons-20210822033320-3217571 ssh "curl -s http://127.0.0.1/ -H 'Host: nginx.example.com'"
    addons_test.go:234: (dbg) Run:  out/minikube -p addons-20210822033320-3217571 addons disable ingress --alsologtostderr -v=1
    addons_test.go:234: (dbg) Done: out/minikube -p addons-20210822033320-3217571 addons disable ingress --alsologtostderr -v=1: (28.727228701s)
=== CONT  TestAddons
    helpers_test.go:184: skipping cleanup of addons-20210822033320-3217571 (--cleanup=false)
--- PASS: TestAddons (122.52s)
    --- PASS: TestAddons/Setup (67.28s)
    --- PASS: TestAddons/parallel (0.00s)
        --- PASS: TestAddons/parallel/Ingress (55.25s)
PASS
Tests completed in 2m2.522589658s (result code 0)
ok      k8s.io/minikube/test/integration        122.560s

##### just-ingress@kvm test:

❯ env TEST_ARGS="-minikube-start-args='--driver=kvm' -test.run TestAddons --cleanup=false" make integration
go test -ldflags="-X k8s.io/minikube/pkg/version.version=v1.22.0 -X k8s.io/minikube/pkg/version.isoVersion=v1.22.0-1628974786-12268 -X k8s.io/minikube/pkg/version.gitCommitID="69d5d34eeb4d5f3ad963c544773dd088d6550821-dirty" -X k8s.io/minikube/pkg/version.storageProvisionerVersion=v5" -v -test.timeout=90m ./test/integration --tags="integration " -minikube-start-args='--driver=kvm' -test.run TestAddons --cleanup=false 2>&1 | tee "./out/testout_69d5d34ee.txt"
Found 16 cores, limiting parallelism with --test.parallel=9
=== RUN   TestAddons
=== RUN   TestAddons/Setup
    addons_test.go:80: (dbg) Run:  out/minikube start -p addons-20210822013905-3004952 --wait=true --memory=4000 --alsologtostderr --driver=kvm --addons=ingress
    addons_test.go:80: (dbg) Done: out/minikube start -p addons-20210822013905-3004952 --wait=true --memory=4000 --alsologtostderr --driver=kvm --addons=ingress: (2m2.902251022s)
=== RUN   TestAddons/parallel
=== RUN   TestAddons/parallel/Ingress
=== PAUSE TestAddons/parallel/Ingress
=== CONT  TestAddons/parallel/Ingress
    addons_test.go:170: (dbg) Run:  kubectl --context addons-20210822013905-3004952 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s
    addons_test.go:170: (dbg) Done: kubectl --context addons-20210822013905-3004952 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s: (10.538721881s)
    addons_test.go:177: (dbg) Run:  kubectl --context addons-20210822013905-3004952 replace --force -f testdata/nginx-ingv1.yaml
    addons_test.go:190: (dbg) Run:  kubectl --context addons-20210822013905-3004952 replace --force -f testdata/nginx-pod-svc.yaml
    addons_test.go:195: (dbg) TestAddons/parallel/Ingress: waiting 4m0s for pods matching "run=nginx" in namespace "default" ...
    helpers_test.go:343: "nginx" [6246bd01-9540-4af8-ab6b-24aad0fdddd8] Pending / Ready:ContainersNotReady (containers with unready status: [nginx]) / ContainersReady:ContainersNotReady (containers with unready status: [nginx])
    helpers_test.go:343: "nginx" [6246bd01-9540-4af8-ab6b-24aad0fdddd8] Running
    addons_test.go:195: (dbg) TestAddons/parallel/Ingress: run=nginx healthy within 15.034188076s
    addons_test.go:215: (dbg) Run:  out/minikube -p addons-20210822013905-3004952 ssh "curl -s http://127.0.0.1/ -H 'Host: nginx.example.com'"
=== CONT  TestAddons
    helpers_test.go:184: skipping cleanup of addons-20210822013905-3004952 (--cleanup=false)
--- PASS: TestAddons (149.03s)
    --- PASS: TestAddons/Setup (122.96s)
    --- PASS: TestAddons/parallel (0.00s)
        --- PASS: TestAddons/parallel/Ingress (26.08s)
PASS
Tests completed in 2m29.033760085s (result code 0)
ok      k8s.io/minikube/test/integration        149.060s

##### full test (filtered):

TestAddons/parallel/Ingress:
13          addons_test.go:170: (dbg) Run:  kubectl --context addons-20210822033320-3217571 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s
14          addons_test.go:170: (dbg) Done: kubectl --context addons-20210822033320-3217571 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s: (13.875082612s)
15          addons_test.go:177: (dbg) Run:  kubectl --context addons-20210822033320-3217571 replace --force -f testdata/nginx-ingv1.yaml
16          addons_test.go:190: (dbg) Run:  kubectl --context addons-20210822033320-3217571 replace --force -f testdata/nginx-pod-svc.yaml
17          addons_test.go:195: (dbg) TestAddons/parallel/Ingress: waiting 4m0s for pods matching "run=nginx" in namespace "default" ...
18          helpers_test.go:343: "nginx" [a150e166-bb61-4d0c-a5d7-bb5d88311075] Pending / Ready:ContainersNotReady (containers with unready status: [nginx]) / ContainersReady:ContainersNotReady (containers with unready status: [nginx])
19          helpers_test.go:343: "nginx" [a150e166-bb61-4d0c-a5d7-bb5d88311075] Running
20          addons_test.go:195: (dbg) TestAddons/parallel/Ingress: run=nginx healthy within 12.015628991s
21          addons_test.go:215: (dbg) Run:  out/minikube -p addons-20210822033320-3217571 ssh "curl -s http://127.0.0.1/ -H 'Host: nginx.example.com'"
22          addons_test.go:234: (dbg) Run:  out/minikube -p addons-20210822033320-3217571 addons disable ingress --alsologtostderr -v=1
23          addons_test.go:234: (dbg) Done: out/minikube -p addons-20210822033320-3217571 addons disable ingress --alsologtostderr -v=1: (28.727228701s)
68          addons_test.go:169: (dbg) Run:  kubectl --context addons-20210822034609-3237971 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s
2708        addons_test.go:169: (dbg) Done: kubectl --context addons-20210822034609-3237971 wait --for=condition=ready --namespace=ingress-nginx pod --selector=app.kubernetes.io/component=controller --timeout=90s: (15.722536526s)
2709        addons_test.go:176: (dbg) Run:  kubectl --context addons-20210822034609-3237971 replace --force -f testdata/nginx-ingv1.yaml
2710        addons_test.go:189: (dbg) Run:  kubectl --context addons-20210822034609-3237971 replace --force -f testdata/nginx-pod-svc.yaml
2711        addons_test.go:194: (dbg) TestAddons/parallel/Ingress: waiting 4m0s for pods matching "run=nginx" in namespace "default" ...
2712        helpers_test.go:343: "nginx" [2053ce1b-96de-452d-ad52-8dd6387d3b38] Pending / Ready:ContainersNotReady (containers with unready status: [nginx]) / ContainersReady:ContainersNotReady (containers with unready status: [nginx])
2742        helpers_test.go:343: "nginx" [2053ce1b-96de-452d-ad52-8dd6387d3b38] Running
2752        addons_test.go:194: (dbg) TestAddons/parallel/Ingress: run=nginx healthy within 14.004000665s
2753        addons_test.go:214: (dbg) Run:  out/minikube -p addons-20210822034609-3237971 ssh "curl -s http://127.0.0.1/ -H 'Host: nginx.example.com'"
2754        addons_test.go:233: (dbg) Run:  out/minikube -p addons-20210822034609-3237971 addons disable ingress --alsologtostderr -v=1
2771        addons_test.go:233: (dbg) Done: out/minikube -p addons-20210822034609-3237971 addons disable ingress --alsologtostderr -v=1: (28.602694275s)

### CODE REVIEW

**Comment by medyagh** [comment]
@prezha thank you can u plz also fix the cilium and other addon failure son docker Linux ?

**Comment by prezha** [comment]
@medyagh i've added changes that should also fix the gcpauth and olm as well

**Comment by medyagh** [comment]
@prezha 

cilium seems to be the last reaminaing test to be fixed then we can merge


https://storage.googleapis.com/minikube-builds/logs/12325/20171/Docker_Linux_containerd.html#fail_TestNetworkPlugins%2fgroup%2fcilium%2fStart

https://storage.googleapis.com/minikube-builds/logs/12325/20171/Docker_Linux.html#fail_TestNetworkPlugins%2fgroup%2fcilium%2fStart

**Comment by medyagh** [comment]
nice job @prezha test look good, just need to fix lint

**Comment by prezha** [comment]
> @prezha
> 
> cilium seems to be the last reaminaing test to be fixed then we can merge
> 
> https://storage.googleapis.com/minikube-builds/logs/12325/20171/Docker_Linux_containerd.html#fail_TestNetworkPlugins%2fgroup%2fcilium%2fStart
> 
> https://storage.googleapis.com/minikube-builds/logs/12325/20171/Docker_Linux.html#fail_TestNetworkPlugins%2fgroup%2fcilium%2fStart

@medyagh sure, cilium is fixed now as well :)

**Comment by sudheerab27** [comment]
@prezha - Can you please share the deploy yaml file for ingress 
regarding - https://github.com/kubernetes/minikube/issues/12299
